src/Functions/ztest.cpp

```diff
@@ -0,0 +1,239 @@
+            data_ci_upper.reserve(input_rows_count);
+
+            auto insert_values = [&data_z_statistic, &data_p_value, &data_ci_lower, &data_ci_upper](Float64 z_stat, Float64 p_value, Float64 lower, Float64 upper)
+            {
+                data_z_statistic.emplace_back(z_stat);
+                data_p_value.emplace_back(p_value);
+                data_ci_lower.emplace_back(lower);
+                data_ci_upper.emplace_back(upper);
+            };
+
+            static const Float64 nan = std::numeric_limits<Float64>::quiet_NaN();
+
+            auto insert_nan_into_columns = [&insert_values]()
+            {
+                return insert_values(nan, nan, nan, nan);
+            };
+
+            boost::math::normal_distribution<> nd(0.0, 1.0);
+
+            for (size_t row_num = 0; row_num < input_rows_count; ++row_num)
+            {
+                const UInt64 successes_x = data_successes_x[row_num];
+                const UInt64 successes_y = data_successes_y[row_num];
+                const UInt64 trials_x = data_trials_x[row_num];
+                const UInt64 trials_y = data_trials_y[row_num];
+                const Float64 confidence_level = data_confidence_level[row_num];
+
+                const Float64 props_x = static_cast<Float64>(successes_x) / trials_x;
+                const Float64 props_y = static_cast<Float64>(successes_y) / trials_y;
+                const Float64 diff = props_x - props_y;
+                const UInt64 trials_total = trials_x + trials_y;
+
+                if (successes_x == 0 || successes_y == 0
+                    || successes_x > trials_x || successes_y > trials_y
+                    || trials_total == 0
+                    || !std::isfinite(confidence_level) || confidence_level < 0.0 || confidence_level > 1.0)
+                {
+                    insert_nan_into_columns();
+                    continue;
+                }
+
+                Float64 se = std::sqrt(props_x * (1.0 - props_x) / trials_x + props_y * (1.0 - props_y) / trials_y);
+
+                /// z-statistics
+                /// z = \frac{ \bar{p_{1}} - \bar{p_{2}} }{ \sqrt{ \frac{ \bar{p_{1}} \left ( 1 - \bar{p_{1}} \right ) }{ n_{1} } \frac{ \bar{p_{2}} \left ( 1 - \bar{p_{2}} \right ) }{ n_{2} } } }
+                Float64 zstat;
+                if (usevar == UNPOOLED)
+                {
+                    zstat = (props_x - props_y) / se;
+                }
+                else if (usevar == POOLED)
+                {
+                    UInt64 successes_total = successes_x + successes_y;
+                    Float64 p_pooled = static_cast<Float64>(successes_total) / trials_total;
+                    Float64 trials_fact = 1.0 / trials_x + 1.0 / trials_y;
+                    zstat = diff / std::sqrt(p_pooled * (1.0 - p_pooled) * trials_fact);
+                }
+                else
+                {
+                    insert_nan_into_columns();
```

Comment:
Probably if use did not pass pooled, or unpooled better throw exception in `getReturnTypeImpl`.